### ISSUE TITLE
Unhandled exception pops up when using Accessibility tools to focus on date button in MonthCalendar

### ISSUE THREAD

**Issue by Olina-Zhang** [dotnet/winforms#7714]
### .NET version

8.0.100-alpha.1.22430.2

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

Yes, it is a regression issue, cannot repro .Net 6.0 and previous, repro in .Net 7.0

### Issue description

Unhandled exception pops up when using Accessibility tools to focus on date button in MonthCalendar

System.InvalidOperationException
  HResult=0x80131509
  Message=Collection was modified after the enumerator was instantiated.
  Source=System.Collections
  StackTrace:
   at System.Collections.Generic.LinkedList`1.Enumerator.MoveNext()
   at System.Windows.Forms.MonthCalendar.CalendarRowAccessibleObject.DisconnectChildren()
   at System.Windows.Forms.MonthCalendar.CalendarBodyAccessibleObject.DisconnectChildren()
   at System.Windows.Forms.MonthCalendar.CalendarAccessibleObject.DisconnectChildren()
   at System.Windows.Forms.MonthCalendar.MonthCalendarAccessibleObject.RebuildAccessibilityTree()
   at System.Windows.Forms.MonthCalendar.MonthCalendarAccessibleObject.OnMonthCalendarStateChanged(Object sender, EventArgs e)
   at System.Windows.Forms.MonthCalendar.WmCalViewChanged(Message& m)
   at System.Windows.Forms.MonthCalendar.WmReflectCommand(Message& m)
   at System.Windows.Forms.MonthCalendar.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)
   at Interop.User32.SendMessageW(IntPtr hWnd, WM Msg, IntPtr wParam, IntPtr lParam)
   at Interop.User32.SendMessageW(IHandle hWnd, WM Msg, IntPtr wParam, IntPtr lParam)
   at System.Windows.Forms.Control.ReflectMessage(IntPtr hWnd, Message& m)
   at System.Windows.Forms.Control.WmNotify(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
   at System.Windows.Forms.Form.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)
   at Interop.User32.CallWindowProcW(IntPtr wndProc, IntPtr hWnd, WM msg, IntPtr wParam, IntPtr lParam)
   at System.Windows.Forms.NativeWindow.DefWndProc(Message& m)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)
   at Interop.User32.DispatchMessageW(MSG& msg)
   at System.Windows.Forms.Application.ComponentManager.Interop.Mso.IMsoComponentManager.FPushMessageLoop(UIntPtr dwComponentID, msoloop uReason, Void* pvLoopData)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(msoloop reason, ApplicationContext context)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(msoloop reason, ApplicationContext context)
   at WinFormsApp8.Program.Main() in C:\Users\xxxx\source\repos\WinFormsApp8\WinFormsApp8\Program.cs:line 14

### Steps to reproduce

**Repro Steps:**

1. Create a Winforms .Net Core application with MonthCalendar added
2. Build and run application
3. Launch Narrator, focus on MonthCalendar, and then click Date button on the top

or use attached test app:
[MonthCalendarSample.zip](https://github.com/dotnet/winforms/files/9469135/MonthCalendarSample.zip)

**Actual result**: Unhandled exception pops up
![MonthCalendarIssue](https://user-images.githubusercontent.com/26474449/187895341-7bc4975a-a6b7-4d5c-89e4-1309ae65669b.gif)

**Comment by vladimir-krestov** [comment]
Sure, #7362 brought this regression. Working on it asap

**Comment by vladimir-krestov** [comment]
@Tanya-Solyanik, do you remember some similar cases, like you pointed for a List collection some time ago?

**Comment by Tanya-Solyanik** [comment]
>@Tanya-Solyanik, do you remember some similar cases, like you pointed for a List collection some time ago?

I see, I missed it here. Could you please review all collections after this one is done?

We also want to service it for 7.0

**Comment by vladimir-krestov** [comment]
I have found the reason. https://github.com/dotnet/winforms/pull/7362 and https://github.com/dotnet/winforms/pull/7569 didn't bring this regression. They exposed the problem in #4758.
The reasons is a double nested call of `RebuildAccessibilityTree` after changing of a view and a range:

![image](https://user-images.githubusercontent.com/49272759/188504402-2bcefdbb-70e9-4dc5-a142-a956e3c1550f.png)

So, the nested call (number 1) releases all elements and makes them null, thereby changing initial states of local fields of the external call (number 2) and we catch the "Collection was modified" exception in the external call because the collections were changed while nested `RebuildAccessibilityTree` call. It happens because we call this methods 2 time at the same time via events. In this case, `CalendarViewChanged` is not necessary here to rebuild the tree because `DisplayRangeChanged` is called as well after view changing, but this event is much universal. So using the both events we rebuild the tree 2 times and have the exception. 

**The solution:** just not use `CalendarViewChanged` event to rebuild the tree because `DisplayRangeChanged` will do it well.

**Comment by Olina-Zhang** [comment]
Verified this issue in the latest .Net 8.0 SDK build: 8.0.100-alpha.1.22469.15, it is fixed:
![7714_Fixed](https://user-images.githubusercontent.com/26474449/191181120-9d16ce3d-2f87-44ac-b390-452c77fa92b3.gif)

**Comment by MandiMan** [comment]
Verified on .NET 7.0 test pass build: .NET 7.0.100-rc.2.22476.6, issue was fixed as below:
![issue7739](https://user-images.githubusercontent.com/108860782/192493449-beb867fd-f864-454d-b631-4daf3b3db446.gif)

**Comment by MandiMan** [comment]
Verified on latest .NET 7 build .NET 7.0.100-rtm.22478.24, issue is fixed, Gif as below:
![issue7739](https://user-images.githubusercontent.com/108860782/193007989-ad745538-501b-40b7-88cd-388e488b8dda.gif)
